magicctapipe/io/io.py

```diff
@@ -0,0 +1,941 @@
+    "load_train_data_files",
+    "load_mc_dl2_data_file",
+    "load_dl2_data_file",
+    "load_irf_files",
+    "save_pandas_data_in_table",
+]
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.StreamHandler())
+logger.setLevel(logging.INFO)
+
+# The telescope IDs and names
+TEL_NAMES = {1: "LST-1", 2: "MAGIC-I", 3: "MAGIC-II"}
+
+# The telescope combination types
+TEL_COMBINATIONS = {
+    "m1_m2": [2, 3],  # combo_type = 0
+    "lst1_m1": [1, 2],  # combo_type = 1
+    "lst1_m2": [1, 3],  # combo_type = 2
+    "lst1_m1_m2": [1, 2, 3],  # combo_type = 3
+}
+
+# The pandas multi index to classify the events simulated by different
+# telescope pointing directions but have the same observation ID
+GROUP_INDEX_TRAIN = ["obs_id", "event_id", "true_alt", "true_az"]
+
+# Event weight for training RFs, but it is set to 1, meaning no weights.
+# ToBeChecked: what weights are best for training RFs?
+EVENT_WEIGHT = 1
+
+# The LST nominal and effective focal lengths
+NOMINAL_FOCLEN_LST = 28 * u.m
+EFFECTIVE_FOCLEN_LST = 29.30565 * u.m
+
+# The upper limit of the trigger time differences of consecutive events,
+# allowed when calculating the ON time and dead time correction factor
+TIME_DIFF_UPLIM = 0.1 * u.s
+
+# The LST-1 and MAGIC readout dead times
+DEAD_TIME_LST = 7.6 * u.us
+DEAD_TIME_MAGIC = 26 * u.us
+
+
+def get_stereo_events(
+    event_data, quality_cuts=None, group_index=["obs_id", "event_id"]
+):
+    """
+    Gets the stereo events surviving specified quality cuts.
+
+    It adds the telescope multiplicity and combination types to the
+    output data frame.
+
+    Parameters
+    ----------
+    event_data: pandas.core.frame.DataFrame
+        Data frame of shower events
+    quality_cuts: str
+        Quality cuts applied to the input data
+    group_index: list
+        Index to group telescope-wise events
+
+    Returns
+    -------
+    event_data_stereo: pandas.core.frame.DataFrame
+        Data frame of the stereo events surviving the quality cuts
+    """
+
+    event_data_stereo = event_data.copy()
+
+    # Apply the quality cuts
+    if quality_cuts is not None:
+        event_data_stereo.query(quality_cuts, inplace=True)
+
+    # Extract stereo events
+    event_data_stereo["multiplicity"] = event_data_stereo.groupby(group_index).size()
+    event_data_stereo.query("multiplicity == [2, 3]", inplace=True)
+
+    # Check the telescope combination types
+    n_events_total = len(event_data_stereo.groupby(group_index).size())
+    logger.info(f"\nIn total {n_events_total} stereo events are found:")
+
+    for combo_type, (tel_combo, tel_ids) in enumerate(TEL_COMBINATIONS.items()):
+
+        multiplicity = len(tel_ids)
+
+        df_events = event_data_stereo.query(
+            f"(tel_id == {tel_ids}) & (multiplicity == {multiplicity})"
+        ).copy()
+
+        df_events["multiplicity"] = df_events.groupby(group_index).size()
+        df_events.query(f"multiplicity == {multiplicity}", inplace=True)
```

Comment:
I do not understand this part. You already select the events by multiplicity in line 114, so why do you recompute multiplicity and apply the cut once again in lines 117-118?

Reply:
Indeed in line 114 we already extracted the events with a given multiplicity, but the problem is that with this condition the events not belonging to a given telescope combination are also extracted due to the condition `f"tel_id == {tel_ids}"`. For example, in case of `tel_ids == [1, 2]`, the condition in line 114 also extracts the events whose telescope multiplicity is 2 but belonging to other combination types (i.e., tel 1 events of the combination [1, 3], or tel 2 events of the combination [2, 3]), because the pandas query handles a list as "or" condition. Thus, it is needed to recalculate the multiplicity and apply the cut in line 117-118 to remove those events, whose multiplicity will be calculated to 1.

Actually I also thought that this part is confusing, but I did not come up with another clear way. So let me add some comments or explanations in the code. If you have some ideas please let me know.

Reply:
ok, I see now. A small explanation in the code would be useful